design-proposals/klip-58-limit-clause-for-pull-queries.md

```diff
@@ -0,0 +1,217 @@
+# KLIP 58 - LIMIT clause for Pull Queries
+
+**Author**: Chittaranjan Prasad (@cprasad1) | 
+**Release Target**: 0.25.0 | 
+**Status**: _In Discussion_ | 
+**Discussion**: _link to the design discussion PR_
+
+**tl;dr:** Users of pull queries don't have any way of restricting the number of rows returned by a pull query. The 
+`LIMIT` clause will enable users to save on computation, time and bandwidth costs by giving them an easy way to restrict 
+the number of rows returned. In the future, the `LIMIT` clause can also be used to write more expressive queries in 
```

Comment:
@AlanConfluent 
>  to request the partitions in some deterministic way rather than all in parallel.

since we don't have an `OFFSET` clause yet in ksqlDB, we can't make any such promises now. We will be requesting the partitions in parallel as part of this KLIP, and won't change that implementation detail right now. We can cover the deterministic behaviour that you mention sometime in the future.
>Whether we fetch partitions in parallel or serially could also be something we expose eventually in EXPLAIN

I will be leaving out discussion about that behavior out of this KLIP for now

Reply:
@mjsax 
> Or are you referring to the sql OFFSET clause? 

yes

> For this case, without ORDER BY it would still not be defined with rows are returned. I think, OFFSET clause is only useful in combination with ORDER BY

I agree. `OFFSET` would only make sense in combination with `ORDER BY`. The `LIMIT` clause proposed in this KLIP is just a stepping stone towards complex queries that include `LIMIT`+`OFFSET`+`ORDER BY`